### PR TITLE
Flesh out typings a little more.

### PR DESCRIPTION
**Pull request by kofi-clever**
Added support for a few more commonly used properties - copied over from https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts

Prepping for adding typescript compilation to clever-components - bringing in stuff that was branched off in apps-dashboard and hubble to support expanded usage of the React API to avoid having to branch again.

- Add `React.[createElement, Children | PureComponent | PropTypes]`
- Add support for optional `context` argument in constructor.
- Better define `setState` to include the optional callback.
- Also removing pre-defined static properties (like propTypes and defaultProps) so clients have to explicitly define them in the component class, encouraging the use of static typing.

Major bump -> v2.0.0

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -6,14 +6,66 @@ export = React;
...
   }
+
+  export class PureComponent extends Component {}
+
   function createElement(...args:any[]): any;
-  var PropTypes : any;
+
+  var Children: ReactChildren;
+  var PropTypes: ReactPropTypes;
+
```
Comment (kofi-clever) [inline index.d.ts:27]: the below defs are copied from @types/react to support `React.Children` and `React.PropTypes` above

File: index.d.ts [hunk 2]
```diff
@@ -6,14 +6,66 @@ export = React;
...
+  //
+  // React.PropTypes
+  // ----------------------------------------------------------------------
```
Comment (arsalansufi) [inline index.d.ts]: nit: Might be a good idea to add where these defs were copied from to this comment

**Review by arsalansufi — APPROVED** [review]
LGTM! Just one nit

File: index.d.ts (hunk 2 above)
Comment (kofi-clever) [reply index.d.ts]: ah good call - done! https://github.com/Clever/minimal-react-typings/pull/3/commits/dd6c0f219d4b1843d5733112c301024b0e6e0f9b

**Pull request merged**